### ISSUE TITLE
Okhttp Non IOException handling for Enqueue

### ISSUE THREAD

**Issue by Tolriq** [square/okhttp#4380]
As I started to implement coroutines with Okhttp I migrated some calls from execute() to enqueue() to have cancellable Coroutines.

So far all is working well, with worker pool and everything, and then comes the non IO related exceptions that where not tested and are not handled by OkHttp at all when using enqueue.

The problem is that the crash occurs in Okhttp threads and so can't be intercepted properly by the application.

```
Caused by android.system.ErrnoException: isConnected failed: ECONNREFUSED (Connection refused)
       at libcore.io.IoBridge.isConnected(IoBridge.java:267)
       at libcore.io.IoBridge.connectErrno(IoBridge.java:187)
       at libcore.io.IoBridge.connect(IoBridge.java:129)
       at java.net.PlainSocketImpl.socketConnect(PlainSocketImpl.java:137)
       at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:390)
       at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:230)
       at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:212)
       at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:436)
       at java.net.Socket.connect(Socket.java:621)
       at okhttp3.internal.platform.AndroidPlatform.connectSocket(AndroidPlatform.java:1)
       at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:81)
       at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:23)
       at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:120)
       at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:15)
       at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:8)
       at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:6)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:12)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:1)
       at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:105)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:12)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:1)
       at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:62)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:12)
       at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:75)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:12)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:1)
       at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:23)
       at okhttp3.RealCall$AsyncCall.execute(RealCall.java:1)
       at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:3)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
       at java.lang.Thread.run(Thread.java:764)
```

This user is probably rooted and have disabled internet permission by default, but application should be able to handle that. And there's a few other errors that are common and Android and require handling.j

This was reported before https://github.com/square/okhttp/issues/3477 and the answer was we can't handle that as it's OS error, but I'm not sure the problem was correctly description in the issue naming and description and that after the user explained more in the issue it was not really analysed after.
(There's a second similar issue https://github.com/square/okhttp/issues/3465 that was close for the more or less same reason).

The fix proposed by the user to add an interceptor works, but when using generic reusable component this means that for each request I need to newbuilder.addinterceptor.build this is not efficient and that I lose the original exception for some exceptions I can parse the message and guess, but having the proper exception is better.

I'd like to request a way to properly handle those directly from OKhttp, I understand that it will be complex as interface default do not exist in Java for Android and that changing the callback exception type would not be binary compatible. But I hope there's a way I did not thought about :)
Errors happens on Android, application can handle them in synchronous ways by catching them and OkHttp should allow the app to get those and do something instead of crashing in place where the app can't intercept.

**Comment by swankjesse** [comment]
Not quite sure I follow. Could you provide the full stacktrace? Maybe an executable test case?

**Comment by Tolriq** [comment]
This is the full stacktrace.
The issue is simple use Call.enqueue() for example
```
enqueue(object : Callback {
                override fun onResponse(call: Call, response: Response) {
                    continuation.resume(response)
                }

                override fun onFailure(call: Call, e: IOException) {
                    // Don't bother with resuming the continuation if it is already cancelled.
                    if (continuation.isCancelled) {
                        return
                    }
                    continuation.resumeWithException(e)
                }
            })
```

One would expect that any error that occurs during the call is intercepted by OkHTTP and transmitted to the onFailure callback.
Unfortunately all exceptions that are not IOExceptions are not catched by OKHttp and will crash the application without a way for the application to catch the problem.

This particular case is generated by user on rooted device that use tools like xPrivacy to by default remove the INTERNET permission from application. Users forget to enable it sometimes and when using OKHttp enqueue() app will crash.

I don't really know how to trigger that specific case by test sorry, or maybe just not add the INTERNET PERMISSION in the manifest is enough if gradle let the application compile.

But this is only one common case, I'm pretty sure with Android 9 new doze buckets there's similar case where internet access will be blocked with some strange errors that may not be IOExceptions. (Not talking about the random errors that can occurs on chinese phones or faulty roms).

TLDR, it would be nice that enqueue() calls does not crash application when a non IOException occurs, as most of the time an asynchronous http call is never vital for the application and the application should easily recover from any kind of non IOException issues.

**Comment by swankjesse** [comment]
So the request is that OkHttp treats all RuntimeExceptions as IOExceptions when calls are asynchronous?

**Comment by Tolriq** [comment]
Not necessarily you are the master here :) (And unfortunately android.system.ErrnoException is a directly child of Exception)

The request is to have a way to have the application not crash in a way that is efficient, and if possible to be aware of the original exception.
Could be a new overload to enqueue that says ignore or convert those to IO, a global unhandled exception to Okhttp, there's certainly many other possibility that I don't think of.

If there's no way to have a solution I suppose I can add the interceptor in the root module that manage the global okhttp and loose the exception for synchronous calls. It's still an unnecessary interceptor that allocate, but less allocation than creating a new builder at each calls.

**Comment by swankjesse** [comment]
Could you paste the entire stacktrace? This “caused by” one is missing information.

**Comment by swankjesse** [comment]
ConnectException is an `IOException` that OkHttp usually delivers to `onFailure()`. If you can show that it’s not, could you provide an executable test case? As-is this isn’t actionable.

**Comment by Tolriq** [comment]
I can't reproduce you have everything I have.

On my device removing INTERNET permission from manifest correctly have the exception handled and forwarded to the correct place in app that then handle it as
```
java.net.SocketException: socket failed: EACCES (Permission denied)
        at java.net.Socket.createImpl(Socket.java:487)
        at java.net.Socket.getImpl(Socket.java:547)
        at java.net.Socket.setSoTimeout(Socket.java:1175)
        at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:243)
        at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:165)
        at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:257)
        at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135)
        at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114)
        at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
        at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
        at com.genimee.android.yatse.mediacenters.kodi.api.KodiApiConnection$getOptionalInterceptor$1.intercept(KodiApiConnection.kt:47)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
        at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
        at okhttp3.RealCall$AsyncCall.execute(RealCall.java:147)
        at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)
```

Since I don't know how to simulate OS Exceptions via test, I suppose this will be as the other reports, categorized as OS issues and that I'll keep the workaround with the interceptor.

I personally prefer when application that start threads that can crash apps, to catch OS errors and forward to let the main application decide what to do, but it's your call.

Typically Chris Banes agreed to catch OS errors in palette async (https://issuetracker.google.com/issues/37047633) despite having an hackish workaround available.

I don't know the cost at execution in performance and allocation of the simplest hack to fix this issue:
```
Interceptor {

        @Throws(IOException::class)
        override fun intercept(chain: Interceptor.Chain): Response {
            try {
                return chain.proceed(chain.request())
            } catch (e: Throwable) {
                if (e is IOException) {
                    throw e
                } else {
                    throw IOException(e)
                }
            }
        }
    }
```
But I fear that it's non null :( so will let you close this issue if you decide that it's a non issue.

**Comment by swankjesse** [comment]
Completely negligible.

Also – the API I think you're looking for is the UncaughtExceptionHandler.

**Comment by Tolriq** [comment]
OkHTTP does not offer an UncaughtExceptionHandler in it's API or can't find it.

If you talk about the global UncaughtExceptionHandler then not really, I do want the app to crash when I made mistake or assume thing that I can handle. Or catch them when I can't and it does not impact the application.

The issue here is that the crash occurs in a place where application can't do anything about it.
For an app that fully support offline mode, it should not crash because there's no route and OS generate a non IO Exception on some devices.

Anyway the interceptor way works, so it's OK for me as cost is not that important when I put it in the main app and not the final module who make the call, so while I still think OkHTTP should handle that I won't fight :)

**Comment by Tolriq** [comment]
@swankjesse I spoke too soon, it seems the interceptor is not 100% viable solution I'm now completely lost.

Just got this crash.
```
Fatal Exception: java.net.SocketTimeoutException: failed to connect to /10.236.185.6 (port 8080) from /192.168.0.132 (port 34928) after 5000ms
       at libcore.io.IoBridge.connectErrno(IoBridge.java:185)
       at libcore.io.IoBridge.connect(IoBridge.java:130)
       at java.net.PlainSocketImpl.socketConnect(PlainSocketImpl.java:129)
       at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:356)
       at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:200)
       at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:182)
       at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:356)
       at java.net.Socket.connect(Socket.java:616)
       at okhttp3.internal.platform.AndroidPlatform.connectSocket(AndroidPlatform.java:73)
       at okhttp3.internal.connection.RealConnection.connectSocket$72690553(RealConnection.java:245)
       at okhttp3.internal.connection.RealConnection.connect$68d0d7c3(RealConnection.java:165)
       at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:257)
       at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135)
       at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114)
       at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
       at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
       at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
       at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
       at com.genimee.android.utils.http.OkHttpWorkerPool$OkHttpExceptionInterceptor.intercept(OkHttpWorkerPool.kt:106)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
       at mediacenters.kodi.api.KodiApiConnection$getOptionalInterceptor$1.intercept(KodiApiConnection.kt:47)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
       at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
       at okhttp3.RealCall$AsyncCall.execute(RealCall.java:147)
       at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
       at java.lang.Thread.run(Thread.java:764)
```

My interceptor is present (OkHttpExceptionInterceptor.intercept) 
```
    private class OkHttpExceptionInterceptor : Interceptor {

        @Throws(IOException::class)
        override fun intercept(chain: Interceptor.Chain): Response {
            try {
                return chain.proceed(chain.request())
            } catch (e: Throwable) {
                if (e is IOException) {
                    throw e
                } else {
                    throw IOException(e)
                }
            }
        }
    }
```

I get it that you do not seems to want to fix at OkHttp, so the only other solution seems to be to use execute like the app does since many, many years.

Is there any fundamental difference or risks in using a volatile call and cancel it from another thread?
Or is the cancelling fully thread safe and will result in mostly the same? (Except manual thread handling of course)

**Comment by swankjesse** [comment]
Please provide an executable test case. I don’t think OkHttp will be able to do more than your interceptor does.

**Comment by Tolriq** [comment]
After discussion with Coroutine lead, probable cause is described at https://github.com/Kotlin/kotlinx.coroutines/issues/830 with some experimental workaround until proper way is made.

I have no idea why the stacktrace is fully the original and not wrapped in anything, but sounds like the only possible explanation.

**Comment by swankjesse** [comment]
Good find. No action to take here in OkHttp so far as I can tell.

**Comment by Tolriq** [comment]
Yep just let people know as that use case is shown in most example of Okhttp and retrofit, I don't know if @JakeWharton retrofit adapter could be touched by that for example.

**Comment by Tolriq** [comment]
So after tons of tests:

https://gist.github.com/Tolriq/dcbdc62c0e29d3b321034e990c3c85ce

A working OkHttp worker pool, that properly handle coroutine race and non IO Exceptions in enqueue.

Could serve for others.

Still think the interceptor to cover https://github.com/square/okhttp/issues/3477 should maybe be native to OkHttp as even if it was not the cause the crashes here, it's still valid.

**Comment by cuichanghao** [comment]
I occur the same error.
is's coroutine's problem.
local scope async cannot catch error.
https://stackoverflow.com/questions/50230466/kotlin-withcontext-vs-async-await

bad code:
'''
launch {
            try {
                val result = async(Dispatchers.IO) { //this line
             }
}
'''


to 

correct code:

'''
launch {
            try {
                val result = withContext(Dispatchers.IO) { //this line
             }
}
'''
